### PR TITLE
new package: py-arrow

### PR DESCRIPTION
**Pull request by healther**
depends on #13567 #13570 #13571

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-arrow/package.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    times and timestamps. It implements and updates the datetime type,
+    plugging gaps in functionality and providing an intelligent module API
+    that supports many common creation scenarios. Simply put, it helps you
+    work with dates and times with fewer imports and a lot less code."""
+
+    homepage = "https://arrow.readthedocs.io/en/latest/"
+    url      = "https://pypi.io/packages/source/a/arrow/arrow-0.14.7.tar.gz"
+
+    version('0.14.7', sha256='67f8be7c0cf420424bc62d8d7dc40b44e4bb2f7b515f9cc2954fb36e35797656')
+
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-arrow/package.py]: Add this:
```python
depends_on('python@2.7:2.8,3.5:', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-arrow/package.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    depends_on('py-setuptools', type='build')
+    depends_on('py-backports-functools_lru_cache@1.5.0', type=('build', 'run'), when='^python@2.7')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-arrow/package.py]:
```suggestion
    depends_on('py-backports-functools_lru_cache@1.2.1:', type=('build', 'run'), when='^python@2.7')
```

File: var/spack/repos/builtin/packages/py-arrow/package.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    depends_on('py-chai@1.1.2', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-arrow/package.py]:
```suggestion
    depends_on('py-chai', type='test')
```

File: var/spack/repos/builtin/packages/py-arrow/package.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    depends_on('py-dateparser@0.7:', type=('build', 'run'))
+    depends_on('py-mock@3.0:', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-arrow/package.py]:
```suggestion
    depends_on('py-mock', type='test')
```

File: var/spack/repos/builtin/packages/py-arrow/package.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    depends_on('py-nose@1.3.7:', type=('build', 'run'))
+    depends_on('py-nose-cov@1.6:', type=('build', 'run'))
+    depends_on('py-pre-commit@1.20:', type=('build', 'run'))
+    depends_on('py-python-dateutil@2.8:', type=('build', 'run'))
+    depends_on('py-pytz@2019.0:', type=('build', 'run'))
+    depends_on('py-simplejson@3.16.0:', type=('build', 'run'))
+    depends_on('py-sphinx@1.8.0:', type=('build', 'run'), when='^python@2.7')
+    depends_on('py-sphinx@2.2.0:', type=('build', 'run'), when='^python@3.5')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-arrow/package.py]: Most of these dependencies aren't actually required, at least in 0.14.7.

File: var/spack/repos/builtin/packages/py-arrow/package.py (hunk 4 above)
Comment (healther) [reply var/spack/repos/builtin/packages/py-arrow/package.py]: also changed the coverage dependencies to test

File: var/spack/repos/builtin/packages/py-arrow/package.py (hunk 2 above)
Comment (healther) [reply var/spack/repos/builtin/packages/py-arrow/package.py]: not according to https://github.com/crsmithdev/arrow/blob/d4afda5d1e9ef7c46f19495fc683974d879cee80/requirements.txt

File: var/spack/repos/builtin/packages/py-arrow/package.py (hunk 1 above)
Comment (healther) [reply var/spack/repos/builtin/packages/py-arrow/package.py]: done

File: var/spack/repos/builtin/packages/py-arrow/package.py (hunk 5 above)
Comment (healther) [reply var/spack/repos/builtin/packages/py-arrow/package.py]: not according to https://github.com/crsmithdev/arrow/blob/d4afda5d1e9ef7c46f19495fc683974d879cee80/requirements.txt
although they have this insane thing that they explicitly specify dependency versions

**Comment by healther** [comment]
general question: How do you get/test the dependencies? Do you have spack stage it? I tend to go to the github pages and look online. The disadvantage there is that in most cases some dependencies are stripped prior to pypi-publication.. I would really love spack to be able to do the parsing directly for me 

sidenode: Should the PythonPackage template contain the 
```
depends_on('python@2.7:2.8,3.3:', type=('build', 'run'))
```
in the same way it currently contains setuptools and `py-foo`?

**Comment by adamjstewart** [comment]
> How do you get/test the dependencies?

See https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#python-dependencies

**Comment by healther** [comment]
> See https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#python-dependencies

Today I learned... Thanks, I should probably have read that some time ago... :)

But I was also asking on a more manual level: Do you `spack stage` the packages? Do you manually download and extract the sources? Do you inspect `setup.py` manually or do you have a script?

**Comment by adamjstewart** [comment]
> But I was also asking on a more manual level: Do you spack stage the packages? Do you manually download and extract the sources? Do you inspect setup.py manually or do you have a script?

No, yes, yes. I download the latest tarball, extract it, and manually inspect the `setup.py` and/or `setup.cfg` for every package.

It's pretty hard to script when it's in the `setup.py` file. Part of the reason setuptools switched to storing the metadata in `setup.cfg` is to make it easier for package managers like us to automate this dependency finding. At some point, we should make something to manually parse `setup.cfg` at least.

File: var/spack/repos/builtin/packages/py-arrow/package.py (hunk 5 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-arrow/package.py]: These dependencies still need to be removed

**Comment by healther** [comment]
dependencies are merged

File: var/spack/repos/builtin/packages/py-arrow/package.py (hunk 5 above)
Comment (healther) [reply var/spack/repos/builtin/packages/py-arrow/package.py]: what about this part from the `setup.cfg`?

```
[tool:isort]
line_length = 88
multi_line_output = 3
include_trailing_comma = true
known_third_party = chai,dateparser,dateutil,mock,pytz,setuptools,simplejson
```

`grep -r . {pytz,simplejson}` reveals that they have imports in their tests, didn't find pre-commit or nose. So I kept the first to as `test` and moved `sphinx` to `doc` type. Not sure if dropping nose is correct

File: var/spack/repos/builtin/packages/py-arrow/package.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    work with dates and times with fewer imports and a lot less code."""
+
+    homepage = "https://arrow.readthedocs.io/en/latest/"
+    url      = "https://pypi.io/packages/source/a/arrow/arrow-0.14.7.tar.gz"
+
+    version('0.14.7', sha256='67f8be7c0cf420424bc62d8d7dc40b44e4bb2f7b515f9cc2954fb36e35797656')
+
+    depends_on('python@2.7:2.8,3.5:', type=('build', 'run'))
+    depends_on('py-setuptools', type='build')
+    depends_on('py-backports-functools_lru_cache@1.5.0:', type=('build', 'run'), when='^python@2.7:2.8')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-arrow/package.py]:
```suggestion
    depends_on('py-backports-functools_lru_cache@1.2.1:', type=('build', 'run'), when='^python@2.7:2.8')
```

File: var/spack/repos/builtin/packages/py-arrow/package.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    depends_on('py-python-dateutil@2.8:', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-arrow/package.py]:
```suggestion
    depends_on('py-python-dateutil', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-arrow/package.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    depends_on('py-chai@1.1.2:', type='test')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-arrow/package.py]:
```suggestion
    depends_on('py-chai', type='test')
```

File: var/spack/repos/builtin/packages/py-arrow/package.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    depends_on('py-mock@3.0:', type='test')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-arrow/package.py]:
```suggestion
    depends_on('py-mock', type='test')
```

File: var/spack/repos/builtin/packages/py-arrow/package.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    depends_on('py-pytz@2019.0:', type='test')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-arrow/package.py]: Not listed, remove

File: var/spack/repos/builtin/packages/py-arrow/package.py [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    depends_on('py-simplejson@3.16.0:', type='test')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-arrow/package.py]: Not listed, remove

File: var/spack/repos/builtin/packages/py-arrow/package.py [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+    depends_on('py-sphinx@1.8.0:1.99.99', type='doc', when='^python@2.7:2.8')
+    depends_on('py-sphinx@2.2.0:', type='doc', when='^python@3.5')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-arrow/package.py]: There is no such thing as a doc dependency.

File: var/spack/repos/builtin/packages/py-arrow/package.py (hunk 2 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-arrow/package.py]: Never trust `requirements.txt`, also things may have changed in develop

File: var/spack/repos/builtin/packages/py-arrow/package.py [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+    depends_on('python@2.7:2.8,3.5:', type=('build', 'run'))
+    depends_on('py-setuptools', type='build')
+    depends_on('py-backports-functools_lru_cache@1.2.1:', type=('build', 'run'), when='^python@2.7:2.8')
+    depends_on('py-python-dateutil', type=('build', 'run'))
+    depends_on('py-chai', type='test')
+    depends_on('py-mock', type='test')
+    depends_on('py-pytz@2019.0:', type='test')
+    # documentation dependencies
+    depends_on('py-sphinx@1.8.0:1.99.99', type='build', when='^python@2.7:2.8')
+    depends_on('py-sphinx@2.2.0:', type='build', when='^python@3.5:')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-arrow/package.py]: Documentation isn't usually built by default. I would either remove the sphinx dependency or add a variant.

**Pull request merged**